src/box/iproto.cc

```diff
@@ -2176,41 +2162,167 @@ iproto_session_push(struct session *session, struct port *port)
+	iproto_thread->join_route[0] =
+		{ tx_process_replication, &iproto_thread->net_pipe };
+	iproto_thread->join_route[1] = { net_end_join, NULL };
+	iproto_thread->subscribe_route[0] =
+		{ tx_process_replication, &iproto_thread->net_pipe };
+	iproto_thread->subscribe_route[1] = { net_end_subscribe, NULL };
+	iproto_thread->error_route[0] =
+		{ tx_reply_iproto_error, &iproto_thread->net_pipe };
+	iproto_thread->error_route[1] = { net_send_error, NULL };
+	iproto_thread->push_route[0] =
+		{ iproto_process_push, &iproto_thread->tx_pipe };
+	iproto_thread->push_route[1] = { tx_end_push, NULL };
+	/* IPROTO_OK */
+	iproto_thread->dml_route[0] = NULL;
+	/* IPROTO_SELECT */
+	iproto_thread->dml_route[1] = iproto_thread->select_route;
+	/* IPROTO_INSERT */
+	iproto_thread->dml_route[2] = iproto_thread->process1_route;
+	/* IPROTO_REPLACE */
+	iproto_thread->dml_route[3] = iproto_thread->process1_route;
+	/* IPROTO_UPDATE */
+	iproto_thread->dml_route[4] = iproto_thread->process1_route;
+	/* IPROTO_DELETE */
+	iproto_thread->dml_route[5] = iproto_thread->process1_route;
+	 /* IPROTO_CALL_16 */
+	iproto_thread->dml_route[6] =  iproto_thread->call_route;
+	/* IPROTO_AUTH */
+	iproto_thread->dml_route[7] = iproto_thread->misc_route;
+	/* IPROTO_EVAL */
+	iproto_thread->dml_route[8] = iproto_thread->call_route;
+	/* IPROTO_UPSERT */
+	iproto_thread->dml_route[9] = iproto_thread->process1_route;
+	/* IPROTO_CALL */
+	iproto_thread->dml_route[10] = iproto_thread->call_route;
+	/* IPROTO_EXECUTE */
+	iproto_thread->dml_route[11] = iproto_thread->sql_route;
+	/* IPROTO_NOP */
+	iproto_thread->dml_route[12] = NULL;
+	/* IPROTO_PREPARE */
+	iproto_thread->dml_route[13] = iproto_thread->sql_route;
+	iproto_thread->connect_route[0] =
+		{ tx_process_connect, &iproto_thread->net_pipe };
+	iproto_thread->connect_route[1] = { net_send_greeting, NULL };
+};
 
+static inline int
+iproto_thread_init(struct iproto_thread *iproto_thread)
+{
+	iproto_thread_init_routes(iproto_thread);
+	slab_cache_create(&iproto_thread->net_slabc, &runtime);
 	/* Init statistics counter */
-	rmean_net = rmean_new(rmean_net_strings, IPROTO_LAST);
-	if (rmean_net == NULL) {
-		slab_cache_destroy(&net_slabc);
-		tnt_raise(OutOfMemory, sizeof(struct rmean),
-			  "rmean", "struct rmean");
+	iproto_thread->rmean = rmean_new(rmean_net_strings, IPROTO_LAST);
+	if (iproto_thread->rmean == NULL) {
+		slab_cache_destroy(&iproto_thread->net_slabc);
+		diag_set(OutOfMemory, sizeof(struct rmean),
+			  "rmean_new", "struct rmean");
+		return -1;
 	}
 
-	if (cord_costart(&net_cord, "iproto", net_cord_f, NULL)) {
-		rmean_delete(rmean_net);
-		slab_cache_destroy(&net_slabc);
-		panic("failed to initialize iproto thread");
-	}
+	return 0;
+}
 
-	/* Create a pipe to "net" thread. */
-	cpipe_create(&net_pipe, "net");
-	cpipe_set_max_input(&net_pipe, iproto_msg_max / 2);
+/** Initialize the iproto subsystem and start network io thread */
+void
+iproto_init(unsigned int threads_count)
```

Comment:
cfg_geti returns int, so better make this function take int param. Also I'd add some check verifying rationality of threads_count value. I mean raise warning/error in case user passes too large value. For instance, I doubt that it makes sense to set value more than 2*cores. I don't insist on this check tho.